weatherman.py

```diff
@@ -0,0 +1,331 @@
+        try:
+            return float(value)
+        except ValueError:
+            return 'NaN'
+
+    def populate_data(self):
+        # get list of files in the folder
+        files = os.listdir(self.folder_path)
+        for file in files:
+            with open(f"{self.folder_path}/{file}", "r") as f:
+                for line in f.readlines()[1:]:
+                    entry = line.strip().split(',')
+                    if entry:
+                        # create a WeatherReading object and append to list
+                        self.weather_reading_list.append(WeatherReading(
+                            entry[DATE_INDEX],
+                            self.safe_float(entry[MAX_TEMP_INDEX]),
+                            self.safe_float(entry[MEAN_TEMP_INDEX]),
+                            self.safe_float(entry[MIN_TEMP_INDEX]),
+                            self.safe_float(entry[MAX_HUMIDITY_INDEX+3]),
+                            self.safe_float(entry[MEAN_HUMIDITY_INDEX+3]),
+                            self.safe_float(entry[MIN_HUMIDITY_INDEX+3])
+                        ))
+
+        return self.weather_reading_list
+
+
+class Calculator:
+    ''' Class to calculate extremes, averages for a given year or month '''
+
+    def __init__(self, weather_reading_list: WeatherReading):
+        self.weather_reading_list = weather_reading_list
+        self.calculation_results = DictDataItem()
+
+    def find_extremes_for_year(self, year):
+
+        # filter readings for the given year
+        year_readings = [
+            reading for reading in self.weather_reading_list if reading.date.startswith(str(year))]
+
+        if year_readings:
+            # find the max, min temperature reading and max humidity reading
+            max_temp_reading = max(
+                year_readings, key=lambda x: x.max_temp if x.max_temp != 'NaN' else float('-inf'))
+            min_temp_reading = min(
+                year_readings, key=lambda x: x.min_temp if x.min_temp != 'NaN' else float('inf'))
+            max_humidity_rating = max(
+                year_readings, key=lambda x: x.max_humidity if x.max_humidity != 'NaN' else float('-inf'))
+            self.calculation_results.add_data('year', year)
+            self.calculation_results.add_data(
+                'max_temp', max_temp_reading.max_temp)
+            self.calculation_results.add_data(
+                'max_temp_date', max_temp_reading.date)
+            self.calculation_results.add_data(
+                'min_temp', min_temp_reading.min_temp)
+            self.calculation_results.add_data(
+                'min_temp_date', min_temp_reading.date)
+            self.calculation_results.add_data(
+                'max_humidity', max_humidity_rating.max_humidity)
+            self.calculation_results.add_data(
+                'max_humidity_date', max_humidity_rating.date)
+
+            return self.calculation_results
+
+        else:
+            return "No data available for this year."
```

Comment:
If you're returning No Data available, do it for all functions for consistency. Although, it would be better if function has a single responsibility and only return single type of value, otherwise we have to keep track of what kind of return value it is